src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs

```diff
@@ -0,0 +1,1306 @@
+            }
+        }
+
+        //Returns the part name in its escaped string form.
+        internal static string GetStringForPartUri(Uri partUri)
+        {
+            Debug.Assert(partUri != null, "Null reference check for this uri parameter should have been made earlier");
+            if (!(partUri is ValidatedPartUri))
+                partUri = ValidatePartUri(partUri);
+
+            return ((ValidatedPartUri)partUri).PartUriString;
+        }
+
+        //This method validates the packUri and returns its two components if they are valid-
+        //1. Package Uri
+        //2. Part Uri
+        internal static void ValidateAndGetPackUriComponents(Uri packUri, out Uri packageUri, out Uri partUri)
+        {
+            //Validate if its not null and is an absolute Uri, has pack:// Scheme.
+            packUri = ValidatePackUri(packUri);
+            packageUri = GetPackageUriComponent(packUri);
+            partUri = GetPartUriComponent(packUri);
+        }
+
+        #endregion Internal Methods
+
+        //------------------------------------------------------
+        //
+        //  Internal Events
+        //
+        //------------------------------------------------------
+        // None
+
+        //------------------------------------------------------
+        //
+        //  Private Methods
+        //
+        //------------------------------------------------------
+
+        #region Private Methods
+
+        /// <summary>
+        /// This method is used to validate the package uri
+        /// </summary>
+        /// <param name="packageUri"></param>
+        /// <returns></returns>
+        private static Uri ValidatePackageUri(Uri packageUri)
+        {
+            if (packageUri == null)
+                throw new ArgumentNullException("packageUri");
+
+            if (!packageUri.IsAbsoluteUri)
+                throw new ArgumentException(SR.UriShouldBeAbsolute);
+
+            return packageUri;
+        }
+
+        //validates is a given uri has pack:// scheme
+        private static Uri ValidatePackUri(Uri packUri)
+        {
+            if (packUri == null)
+                throw new ArgumentNullException("packUri");
+
+            if (!packUri.IsAbsoluteUri)
+                throw new ArgumentException(SR.UriShouldBeAbsolute);
+
+            if (packUri.Scheme != PackUriHelper.UriSchemePack)
+                throw new ArgumentException(SR.UriShouldBePackScheme);
+
+            return packUri;
+        }
+        
+        private static readonly char[] HexUpperChars = {
+                                   '0', '1', '2', '3', '4', '5', '6', '7',
+                                   '8', '9', 'A', 'B', 'C', 'D', 'E', 'F' };
+        
+        private static string HexEscape(char character) {
+            if (character > '\xff') {
+                throw new ArgumentOutOfRangeException("character");
+            }
+            char[] chars = new char[3];
```

Comment:
I have no idea if this function is perf critical or not, but the chars array could be stackalloc'd instead of heap allocated.